### DISCUSSION TITLE
Unit Test to Validate Metric Collection

### DISCUSSION THREAD

**Discussion by JohnYoungers** [open-telemetry/opentelemetry-dotnet#5637] [Q&A] [answered]
Given I'm testing a aspnetcore application using `WebApplicationFactory<TStartup>`, what would be the easiest way to get the list of all metrics that are being tracked?  Is there something in the server's `IServiceProvider` I could pull? Or maybe setup an endpoint or custom exporter?

The aspnetcore application being tested isn't an actual API, just one to verify various functionality, so exposing endpoints or any other functionality isn't a concern.

As we upgrade our OTEL packages, I'd like to verify which metrics are added/renamed, etc, so we can better monitor data ingestion levels with releases

**Comment by vishweshbankwar** [comment, marked as answer]
@JohnYoungers You could use [InMemoryExporter](https://www.nuget.org/packages/OpenTelemetry.Exporter.InMemory) to collect the metrics for verification. e.g. https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs.

**Reply by JohnYoungers** [reply]
thank you: I'll give it a shot.

I had tried this approach earlier (adding the list as a singleton in the services, and adding the `InMemoryExporter`), but the list was always empty in the tests; I'll check again to see what I was missing

**Reply by JohnYoungers** [reply]
It looks like if I grab the list and then dispose of the `WebApplicationFactory`, it will be populated. Thank you!

**Reply by mu88** [reply]
@vishweshbankwar: I have the same question, too, and you're approach works nice. However, actually running an in-memory ASP.NET Core app feels like an integration test to me.
Do you also see a possibility to write a unit test based on the `IHostApplicationBuilder` enriched by OpenTelemetry? I want to write unit tests for my extension method (see https://github.com/mu88/mu88.Shared/blob/main/src/mu88.Shared/OpenTelemetry/HostApplicationBuilderExtensions.cs#L40), however, a lot of the OpenTelemetry building blocks are `internal`. I'd really appreciate it if you could give me a good starting point for this without me reverse engineering OpenTelemetry 😄
